### PR TITLE
Modify SpeechSynthesisDataset class, make it return text

### PR DESCRIPTION
**Pull request by yaozengwei**
This PR is required by the TTS recipe https://github.com/k2-fsa/icefall/pull/1372 in icefall. In this TTS recipe, we convert the transcript text to phonemes in training.

### CODE REVIEW

File: lhotse/dataset/speech_synthesis.py [hunk 1]
```diff
@@ -19,26 +19,35 @@ class SpeechSynthesisDataset(torch.utils.data.Dataset):
...
+            'tokens_lens': (B, ) int tensor  # when return_tokens=True
+            'text': List[str] of len B  # when return_tokens=False
+            'cut': List of Cuts  # when return_cuts=True
         }
     """
 
     def __init__(
         self,
-        cuts: CutSet,
+        return_cuts: bool = False,
```
Comment (pzelasko) [inline lhotse/dataset/speech_synthesis.py:33]: This change is breaking, can you move `return_cuts` to some position towards the end of parameter list?

**Review by pzelasko — COMMENTED** [review]
Looks good, but don't you prefer to have G2P inside `SpeechSynthesisDataset` to avoid spending the time running it inside the training loop?

File: lhotse/dataset/speech_synthesis.py (hunk 1 above)
Comment (yaozengwei) [reply lhotse/dataset/speech_synthesis.py:33]: OK.

**Comment by yaozengwei** [comment]
> Looks good, but don't you prefer to have G2P inside `SpeechSynthesisDataset` to avoid spending the time running it inside the training loop?

Hi, @pzelasko,  I prefer to do the text normalization and tokenization in separate training recipes, since they usually depend on different packages. @csukuangfj suggests doing this in data preparation stage.

**Comment by pzelasko** [comment]
> > Looks good, but don't you prefer to have G2P inside `SpeechSynthesisDataset` to avoid spending the time running it inside the training loop?
> 
> Hi, @pzelasko, I prefer to do the text normalization and tokenization in separate training recipes, since they usually depend on different packages. @csukuangfj suggests doing this in data preparation stage, with converted phonemes saved to manifests.

OK cool. It looks like there are some conflicts after merging the other PR with multi-speaker support, could you resolve them?

**Comment by yaozengwei** [comment]
> > > Looks good, but don't you prefer to have G2P inside `SpeechSynthesisDataset` to avoid spending the time running it inside the training loop?
> > 
> > 
> > Hi, @pzelasko, I prefer to do the text normalization and tokenization in separate training recipes, since they usually depend on different packages. @csukuangfj suggests doing this in data preparation stage, with converted phonemes saved to manifests.
> 
> OK cool. It looks like there are some conflicts after merging the other PR with multi-speaker support, could you resolve them?

Ok. Thanks. I have some local changes and will resolve the conflicts later. 

In addition, in current implementation, it will load the whole cuts set and generate a char-based vocabulary according to the given texts. But I think TTS recipes usually use phoneme tokens instead. Also, if might cause token-id-mapping mismatch bettwen training and test, when separately given two cuts sets for this class. Shall we remove this and make this calss return raw text and optionally pre-converted (phoneme) tokens? 

https://github.com/lhotse-speech/lhotse/blob/706117511588bf736601e034dc052e0426cbb3b0/lhotse/dataset/speech_synthesis.py#L43

**Comment by pzelasko** [comment]
> In addition, in current implementation, it will load the whole cuts set and generate a char-based vocabulary according to the given texts. But I think TTS recipes usually use phoneme tokens instead. Also, if might cause token-id-mapping mismatch bettwen training and test, when separately given two cuts sets for this class. Shall we remove this and make this calss return raw text and optionally pre-converted (phoneme) tokens?

I see that this class accepts `cuts` in constructor, it's an outdated design we don't use anymore, so thanks for updating it. Instead, it would make sense to pass a `tokenizer` object/callable that (optionally) converts raw text into some tokens. What you're suggesting also works.

**Comment by yaozengwei** [comment]
@pzelasko Thanks. It is ready to be merged.

**Comment by yaozengwei** [comment]
@pzelasko @desh2608 Would you mind giving a review if you have time?
